examples/OcclusionSaliency.ipynb

```diff
@@ -1,151 +1,301 @@
+    }
+   ],
+   "source": [
+    "import PIL.Image\n",
+    "\n",
+    "test_image_filename = 'catdog.png'\n",
+    "PIL.Image.open(test_image_filename)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# The \"application\"\n",
+    "First we will set up our application as a function with defined inputs for\n",
+    "  * the image\n",
+    "  * black-box classifier\n",
+    "  * image perturbation API implementation\n",
+    "  * saliency map generation API implementation\n",
     "\n",
+    "For the purposes of this tool, let's say that the input blackbox classifier must take in one image and output a 1D vector of per-class confidences (`Callable[[PIL.Image.Image], np.ndarray]`) for simplicity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
     "import matplotlib.pyplot as plt\n",
-    "from PIL import Image\n",
-    "from torchvision import transforms\n",
-    "from torch.autograd import Variable\n",
-    "import torchvision.models as models\n",
-    "import torch\n",
     "import numpy as np\n",
-    "from operator import itemgetter\n",
-    "import os\n",
+    "import PIL.Image\n",
+    "from typing import Callable\n",
+    "from xaitk_saliency import PerturbImage, ImageClassifierSaliencyMapGenerator\n",
+    "\n",
+    "\n",
+    "def app(\n",
+    "    image_filepath: str,\n",
+    "    # Assuming outputs `nClass` length arrays.\n",
+    "    blackbox_classify: Callable[[PIL.Image.Image], np.ndarray],\n",
+    "    perturber: PerturbImage,\n",
+    "    sal_map_generator: ImageClassifierSaliencyMapGenerator,\n",
+    "):\n",
+    "    # Load the image\n",
+    "    ref_image = PIL.Image.open(image_filepath)\n",
+    "    # Use the perterbation API implementation input to generate a bunch of images.\n",
+    "    pert_imgs, pert_masks = perturber(ref_image)\n",
+    "    print(f\"Perterbed images: {len(pert_imgs)}\")\n",
+    "    print(f\"Perterbation masks: {pert_masks.shape}\")\n",
+    "    # For the saliency heatmap generation API we need reference image predictions as well as\n",
+    "    # the predictions for each of the perturbed images.\n",
+    "    ref_preds = blackbox_classify(ref_image)\n",
+    "    print(f\"Ref preds: {ref_preds.shape}\")\n",
+    "    pert_preds = np.asarray([\n",
+    "        blackbox_classify(pi)\n",
+    "        for pi in pert_imgs\n",
+    "    ])\n",
+    "    print(f\"Pert preds: {pert_preds.shape}\")\n",
+    "    sal_maps = sal_map_generator(ref_preds, pert_preds, pert_masks)\n",
+    "    print(f\"Saliency maps: {sal_maps.shape}\")\n",
+    "    \n",
+    "    # Visualize mask image region coverage\n",
+    "    mask_coverage = np.zeros(pert_masks[0].shape)\n",
+    "    for pm in pert_masks:\n",
+    "        mask_coverage += pm\n",
+    "    mask_coverage /= len(pert_masks)\n",
+    "\n",
+    "    plt.figure(figsize=(4, 4))\n",
+    "    plt.imshow(mask_coverage)\n",
```

Comment:
Maybe don't use autoscaling here when showing mask coverage? It amplifies very small differences when it should be relatively uniform for sliding window and RISE.

Also- I think for RISE specifically, we need to divide mask_coverage by the expectation (i.e. p=0.5). That will make the mask coverage values closer to 1, which corresponds to the sliding window approach.

Reply:
The application is intentionally not implementation specific. Maybe a "solution" is we not have that mask region coverage part in there at all as it seems to be a confusor to the point of the example. Additionally, is it maybe the case that the differences are appropriate? E.g. regions are much more likely to be at least partially perturbed in RISE than the sliding-window method.